### ISSUE TITLE
More powerful alternative sentence template system

### ISSUE THREAD

**Issue by chickendude** [kantord/LibreLingo#1122]
### The issue
At the moment with simpler sentences (The X is Y) it's not a huge issue, but as sentences get more complex there will be more and more valid alternatives. Entering each alternative entry by hand when they number in the hundreds will be a chore, and will need to be done for each sentence. Take for example the following simple Spanish sentence:

> Lo vi hace poco.

I saw him a short while ago.
I saw it a short while ago.
I saw him not long ago.
I saw it not long ago.
I saw him a little while ago.
I saw it a little while ago.
etc.

### A possible solution

We can use a template system which lets you create alternative values within a sentence. For example, the previous sentences could be constructed as such:
```
I saw [him/it] a [short/little] [while/bit] [ago/back].
I saw [him/it] [not long ago/recently].
```

You can also use an empty value before/after the "/", i.e. "[/...", to indicate nothing:
```
I saw [him/it] not [/that] long ago.
```
Gives you "I saw him/it not long ago" and "I saw him/it not that long ago".

Ideally we would be able to test these (and perhaps save the tests to ensure future sentences don't break them) to ensure that incorrect responses aren't accepted.

### Conclusion

This should help remove a lot of boilerplate and copying/pasting and facilitate the process of supporting a wider range of acceptable alternative answers.

**Comment by kantord** [comment]
This sounds very useful!

My approach to this is to split this into 2 steps:

- Find a simple command line software that can already do this. If it doesn't exist, create it
- Set up the YAML loader so that it can run a series of command line programs to filter phrases through

Essentially a phrase filter would be a function/command line program that can take a list of phrases and convert it to another list of phrases, even if it's of a different length. 

Another example of a possible filter would be a filter that would automatically create variants of words even without the need to specify it in variants. That could be useful for supporting different spelling variants such as "color" vs "colour"

So essentially I really like the idea, but instead of complicating the template system, I would add "filters" to the template system so that additional functionality/hacks can be done through external plugins, which would basically make the templating system way more configurable and extensible.

**Comment by kantord** [comment]
Yes another example of such a filter could be a simple `sed` command. In your example above,

```
I saw [him/it/you] not [/that] long ago.
```

would result in this phrase:

```
I saw you not  long ago.
```

but the triple space could be automatically replaced with a single one using `sed "s/ +/ /g"`

**Comment by chickendude** [comment]
Yeah, some cleaning up of the final string would be helpful regardless (same with input strings, as well as making punctuation optional).

**Comment by kantord** [comment]
> (same with input strings, as well as making punctuation optional).

that's a different issue, mapping of answers is handled by this package: https://github.com/kantord/LibreLingo/blob/main/apps/answer-corrector/src/index.ts

but it should already ignore punctuation! Or more specifically, it should accept your answer but point out that your punctuation was wrong

**Comment by chickendude** [comment]
I've been going through @RGBradley 's branch proofreading the Basque course and having this would be a huge help, i think for the time being we shouldn't worry about adding alternative translations until something like this has been implemented. E.g. this is what i added:
>   - Phrase: Epa. Ni ere irakasle bat naiz.
    Alternative versions:
     - Epa. Ni ere irakaslea naiz.
     - Epa. Neu ere irakasle bat naiz.
     - Epa. Neu ere irakaslea naiz.
     - Kaixo. Ni ere irakasle bat naiz.
     - Kaixo. Ni ere irakaslea naiz.
     - Kaixo. Neu ere irakasle bat naiz.
     - Kaixo. Neu ere irakaslea naiz.

And this is only a small selection of the possible answers, there are several other combinations that could be added, with the template system it would be a one-liner ;)

What sort of commandline software are you looking for? A Python script? I could look into writing a C program or Python script that handles this, perhaps puts the output in a separate file? How do you imagine the inputs/outputs being? I think it would just need to be run once and the expanded list of sentences could be cached/stored somewhere so they don't need to be created each time a user starts a new study session.

**Comment by zmbc** [comment]
@kantord This is a big challenge I'm running into with the toki pona course. Even more so than other languages, toki pona phrases will have many, many possible English translations and they increase exponentially with sentence complexity, as in the examples above.

Personally, it doesn't make sense to me why a command line program is the best way to approach this. Unless there is an existing POSIX command or similar which does this (which I am fairly sure there is not -- I do not understand how `sed` could be used here), that seems like it would just add a layer or indirection and an external dependency.

I'd be interested in implementing this, if it's alright for it not to be a command line program. However, I have two major design questions:

- Is it a good idea to expand these in the course compilation step? You wouldn't have to add too many branches to generate a lot of alternatives. The other approach would be to allow the templating syntax to pass all the way through, and be used by the frontend to match to the submitted answer. What do you think?
- Should nested brackets be allowed? I tend to think yes.

**Comment by zmbc** [comment]
By the way, and maybe this should be moved to a new issue, but there is an even bigger challenge that is closely related.

The way I think about LibreLingo "alternatives" is that they are the versions that are correct, but not taught. The system never prompts the user with them, but it accepts them if the user already knows them.

Sometimes, as a course author, you want to teach **multiple translations** for a single source language phrase. Or sometimes, you may want to teach **multiple versions** in the source language that all translate to the same target language phrase.

The only way to do this in LibreLingo is to copy and paste, thereby creating duplication and the ability for things to get out of sync. If complex alternatives are involved, that amount of duplication can be a real problem.

The terminology used by LibreLingo is also confusing in this regard. There are `Phrase`, `Alternative versions`, `Translation`, and `Alternative translations`. But the thing that one might not realize right away, given the terminology, is that **`Alternative versions` are really a property of `Translation`, and `Alternative translations` are really a property of `Phrase`**. `Alternative versions` holds all valid source -> target language translations of `Translation`. `Alternative translations` holds all valid target -> source translations of `Phrase`.

For this reason, the current workflow is this: If I want to add a second *taught* phrase in the target language, I copy-and-paste the entire item, then change `Phrase` and `Alternative translations`. If I want to add a second *taught* phrase in the source language, I copy-and-paste the entire item, then change `Translation` and `Alternative versions`.

One could imagine greatly increasing the number of exercises available for practice by having N source-language phrases and M target-language phrases, where any can be matched with any in the opposite language. Right now, this would require the author to copy and paste N x M times. There would be N values of `Alternative translations`, and each would be duplicated M times. There would be M values of `Alternative versions`, and each would be duplicated N times.

For my specific case of toki pona, N is always 1. But even just dealing with M scaling up is a pretty big hassle.

**Comment by zmbc** [comment]
Just as a thought experiment, here is one idea for a YAML structure that would avoid all this duplication:

```yaml
Phrase exercises:
  Source:
     - phrase 1 in source language
     - phrase 2 in source language
   Target:
     - phrase 1 in target language
     - phrase 2 in target language
Alternatives:
  - Phrase: phrase in source language 1
    Language: source
    Alternative translations:
      - another translation
      - another translation
  - Phrase: phrase in target language 1
    Language: target
    Alternative translations:
      - another translation
      - another translation
...
```

... which would generate four pairs of phrases for creating exercises, one from each combination.

**Comment by zmbc** [comment]
Also perhaps of interest: [some old documentation of how Duolingo implements all of this](https://incubator.duolingo.com/courses/en/de/help).

According to [this post](https://forum.duolingo.com/comment/38731989/New-browser-extension-for-viewing-the-lists-of-accepted-solutions), there are Duolingo sentences with thousands of acceptable translations -- which definitely makes me think implementing this in a compilation step might not be a good idea.

**Comment by kantord** [comment]
> According to this post, there are Duolingo sentences with thousands of acceptable translations -- which definitely makes me think implementing this in a compilation step might not be a good idea.

why? 

> Is it a good idea to expand these in the course compilation step? You wouldn't have to add too many branches to generate a lot of alternatives. The other approach would be to allow the templating syntax to pass all the way through, and be used by the frontend to match to the submitted answer. What do you think?

If possible then yes, in my opinion it would be best to put it into the compilation step. Actually, I wouldn't even put it into the compilation step. I would put it into the step that reads the YAML files. The reason I'd do this is that in my opinion it's best to do it as early as possible, which is the only easy way to make totally sure that the implementation works the same on every possible client. There are people already developing alternative clients for LibreLingo, meaning if we implement it like you say, they'll have to implement all of this just to support existing courses. I think the purpose of LibreLingo's infrastructure should be to allow anyone to use the course content in a creative way with the least amount of hassle.

**Comment by kantord** [comment]
> Should nested brackets be allowed? I tend to think yes.

can you show me an example of nested brackets?

**Comment by kantord** [comment]
> The only way to do this in LibreLingo is to copy and paste, thereby creating duplication and the ability for things to get out of sync. If complex alternatives are involved, that amount of duplication can be a real problem.

This is true, however for me it raises the question: why would you actively prompt multiple versions of the same sentence, when you could instead prompt different sentences? The user will face the same sentence several times already, adding even more repetition seems like a bad idea. I'd rather prefer to create a similar sentence but with some words changed up

**Comment by kantord** [comment]
> Just as a thought experiment, here is one idea for a YAML structure that would avoid all this duplication:
> 
> ```yaml
> Phrase exercises:
>   Source:
>      - phrase 1 in source language
>      - phrase 2 in source language
>    Target:
>      - phrase 1 in target language
>      - phrase 2 in target language
> Also accepted:
>   - Phrase: phrase in source language 1
>     Language: source
>     Translations:
>       - another translation
>       - another translation
>   - Phrase: # If an array, all members have the same set of translations accepted.
>       - phrase in target language 1
>       - phrase in target language 2
>     Language: target
>     Translations:
>       - another translation
>       - another translation
> ...
> ```
> 
> ... which would generate four pairs of phrases for creating exercises, one from each combination.
> 
> #1619 could use the same system to specify alternate versions of a phrase "translated" from its audio, perhaps using a `Transcriptions` key alongside `Translations`, or generalizing (and renaming) the `Language` key to include "source", "target", and "audio."

if we are now proposing changes to the YAML structure, then why not rather use something like this: 

```yaml
Phrases:
  - English:
      - Lorem ipsum dolor sit amet
      - Lorem ipsum dolor sit amet
    Spanish: Lorem ipsum dolor sit amet
```

**Comment by kantord** [comment]
> I'd be interested in implementing this, if it's alright for it not to be a command line program.

Now I tend to agree with this one in particular, I would implement it specially as part of the YAML loader in Python

**Comment by zmbc** [comment]
> If possible then yes, in my opinion it would be best to put it into the compilation step. Actually, I wouldn't even put it into the compilation step. I would put it into the step that reads the YAML files. The reason I'd do this is that in my opinion it's best to do it as early as possible, which is the only easy way to make totally sure that the implementation works the same on every possible client. There are people already developing alternative clients for LibreLingo, meaning if we implement it like you say, they'll have to implement all of this just to support existing courses.

This is a good point and not something I was aware of. I still have major reservations about it, which I'll explain in a second, but if you still think this outweighs them then I understand.

> > According to this post, there are Duolingo sentences with thousands of acceptable translations -- which definitely makes me think implementing this in a compilation step might not be a good idea.
>
> why?

Data size. Because Duolingo has the templating syntax described above implemented in the frontend, thousands is a low estimate for how many there would be if we enumerated them all.

toki pona is likely the worst possible scenario in this way -- sentences in toki pona have many, many English equivalents. To see exactly how many, I chose a medium-length, pretty average toki pona sentence:

*ona li toki ala e ni: mi pali e tomo suli*

The English equivalent template is quite complex:

*[[He/She/It] [[doesn't/does not/did not/didn't/will not/won't] [say/think]/[isn't/is not/was not/wasn't/will not be] [saying/thinking]]/They [[do not/don't/did not/didn't/will not/won't] [say/think]/[aren't/are not/weren't/were not/will not be] [saying/thinking]]] [this:/that] [[I/we] [make/made/will make/work on/worked on/will work on/build/built/will build/construct/constructed/will construct]/[I [am/will be/was]/We [are/will be/were]] [making/working on/building/constructing]] [[a/the] [big/large/important/old] [house/room/structure]/[the/] [big/large/important/old] [houses/rooms/structures]]*

Which results in **405,504** combinations. With each being about 40 characters, and assuming 10 phrases per skill and 30 skills in the course, the toki pona course would be at least **4.9 gigabytes**. Each skill the user navigated to would require them to download roughly a 200MB file.

Is this acceptable? I don't know. Maybe it is. But that is why I was hesitant. And also, I suspect, why Duolingo does a good portion of the grading on the client.

**Comment by zmbc** [comment]
> if we are now proposing changes to the YAML structure, then why not rather use something like this:

I think we're on the same page -- that is what I meant in my proposal. "Target" and "source" were just placeholders that would be replaced by the actual language names.

This could also potentially be a better way to organize `Also accepted`:

```yaml
Also accepted:
  Spanish:
    - Phrase: phrase in Spanish 1
      Translations:
        - another translation
        - another translation
  English:
    - Phrase: # If an array, all members have the same set of translations accepted.
        - phrase in English 1
        - phrase in English 2
      Translations:
        - another translation
        - another translation
```

**Comment by zmbc** [comment]
> can you show me an example of nested brackets?

[This [would/could] be/Here is] an example of nested brackets.

> This is true, however for me it raises the question: why would you actively prompt multiple versions of the same sentence, when you could instead prompt different sentences?

There are two reasons I want to do it:

- The ambiguity of toki pona is one of its most defining features, which means I want to explicitly draw attention to it **(this reason is very language-specific)**
- Honestly, it is a lot of work to come up with sentences, their alternatives, etc. Even if it would be better to have two distinct sentences, it is probably better to have two exercises with the same sentence translating to different things than to have fewer exercises. **(this reason is likely common to most languages)**

I definitely understand your point here though and I get why this might be a wontfix or not high-priority. I can certainly manage with copy-and-pasting a lot, it's not the end of the world. The template syntax, on the other hand, is absolutely necessary -- I cannot complete the toki pona course without it.

**Comment by zmbc** [comment]
Also, one addendum to this:

> Is this acceptable? I don't know. Maybe it is. But that is why I was hesitant. And also, I suspect, why Duolingo does a good portion of the grading on the client.

If the network usage is not acceptable, I think that it is a non-trivial CS problem to figure out an approach that is good on both runtime and file size (we haven't mentioned runtime yet in this discussion, but it will be an issue if it's not well thought out). I would be willing to work on this problem, if you agree that it is needed.

**Comment by kantord** [comment]
> Each skill the user navigated to would require them to download roughly a 200MB file.
>
> Is this acceptable? I don't know.

It's definitely not acceptable 😅

**Comment by kantord** [comment]
> [This [would/could] be/Here is] an example of nested brackets.

Do you think these could be avoided? I have some concerns:

- It's harder to implement a parser for this, which could result in
  * different clients being out of sync
  * extra dependencies (parser tooling) being added to clients 
- If I remember correctly, brackets are also now used to create multi-word chips for chips challenges. While we could say that only the first layer of brackets will have this feature with nested brackets, usage could be confusing
- Sounds like sentences that would need nested brackets are harder to deal with
  * those could be time sinks for course authors. It's better if they get frustrated with a sentence that's a time sink and rather remove it in favor of a better sentence
  * it would be hard to read these, making mistakes harder to spot or correct
- We'll eventually want to introduce a GUI for editing these, if we make this too complex then the GUI could become too messy

Actually these also make me think: we'll need automated tooling (bots?) for improving the formatting of course content, i. e. automatically introducing brackets for example if a course author writes: "foo bar baz" and "foo lorem baz", there should be a bot that will automatically change that to "foo [bar/lorem] baz" 

Let me know how you feel about these! And, what's your expected timeline for rolling out this feature and how much does it block you right now with the toki pona course?

I was thinking that since these features seem to have a great effect on any potential design for GUI editors it would be nice to at least partially design the editors at the same time of designing this bracket system that basically amounts to a template language 🤔 or better yet, even start implementing the GUI so that we can make small steps and everything is in sync. But I also want to be aware how this affects the development of existing courses.

From my side, I want to introduce free-form typing challenges for phrases, not just words, but delaying that is also one potential way to win time to implement some of the tooling around this, as I imagine words have a much less need for this than phrases do

**Comment by kantord** [comment]
> The ambiguity of toki pona is one of its most defining features, which means I want to explicitly draw attention to it (this reason is very language-specific)

I recognize language specific needs and differences is teaching approach and thus I don't want to be a barrier in this, though I also see the potential dangers of this approach. I guess the only way to figure out is to do extensive testing. 

The way I imagine something like this could be implemented is basically introducing a setting as to how many alternatives should result in challenges being generated. I imagine something like 1 + x% of alternatives should be used as "main" phrases. And x would be 0 by default, therefore this would be disabled, but for toki pona or some other languages the setting could be set to say x = 20, so 1 + 20% of alternatives would result in challenges being prompted to the user.

**Comment by kantord** [comment]
> Honestly, it is a lot of work to come up with sentences, their alternatives, etc. Even if it would be better to have two distinct sentences, it is probably better to have two exercises with the same sentence translating to different things than to have fewer exercises. (this reason is likely common to most languages)

Regarding this one, I'm not 100% sure what you mean. Can you show me an example?

**Comment by kantord** [comment]
> I can certainly manage with copy-and-pasting a lot, it's not the end of the world

It's certainly important to keep in mind that when a lot of copy-and-pasting happens it seems to be a sign of something being wrong and it can potentially create maintenance issues in the future

**Comment by zmbc** [comment]
> Do you think these could be avoided? I have some concerns:

Yes, I also have some concerns. I'm going to need to think about this more, and come up with an algorithmic approach to propose. Then we can discuss from there.

I don't think the GUI has to be overly complex, no matter what we do: even on Duolingo, users just edit these as strings, whether they have template syntax in them or not.

I do have one question that can help inform my brainstorm of an algorithm. You mentioned that people are writing other clients and we don't want to make this overly complicated for them. Do you think it would be helpful to implement the algorithm, maybe in a few languages, and offer it as a package/module that they can use in their clients?

**Comment by kantord** [comment]
> Do you think it would be helpful to implement the algorithm, maybe in a few languages, and offer it as a package/module that they can use in their clients?

Yes, this is one approach I was considering. Actually we kinda already have such a package, it's just much simpler: https://github.com/kantord/LibreLingo/blob/main/apps/answer-corrector/src/index.ts

Another thing that we can do is pre-parse the template syntax and send it as a JSON structure.

So, say, sth like:

> [Lorem/bar baz] [ipsum/dolor sit/foo] amet

could be sent to the clients in this format:

```json
{
  "template": [
    ["lorem", "bar baz"], ["ipsum", "dolor sit", "foo"], "amet"
  ]
}
```

then, this structure could be used to validate a sentence without knowing how to parse the template syntax.

Then one question still remains, how do you validate the sentences on client side? I can think of 

- Generate all possibilities, put them in a set/hash map structure and decide if the user response is part of it (high memory usage and/or slower performance)
- Construct some form of parser based on the template (basically the template is a grammar description) and use that parser to decide whether the sentence is correct. This is much more optimal but much harder to implement, and raises the question: how to decide when the user made a typo?

Actually, both of them raise the question: how to decide when a user made a typo and what that typo is?

Currently this is used:

https://github.com/kantord/LibreLingo/blob/87f7f85964216333beb8a7597de8607d7a2f517c/apps/answer-corrector/src/index.ts#L12-L13

One thing I can imagine is that we generate all possible solutions on the fly as a generator, so never store all of them in memory, we simply iterate over the possibilities. This means we might have to do this multiple times, but we won't increase the memory usage. It's probably worth a try testing if this is even feasible to do in the browser in terms of performance, for example is it possible to iterate over thousands of possible solutions in say, less than 50ms or so

**Comment by zmbc** [comment]
> I imagine something like 1 + x% of alternatives should be used as "main" phrases.

This won't work. As I mentioned above, every "main"/taught phrase needs to have a full set of translations in the opposite language. This means you can't automatically promote an alternative to be a main phrase unless you require all *alternatives* to have full sets of translations, which would be massive amounts of work.

Furthermore, once templating syntax is added, alternatives will have significantly different behavior than phrases/translations and interchanging between the two would become completely impossible.

I still think my YAML proposal above is a good start on how to allow this kind of thing without requiring it.

> Regarding this one, I'm not 100% sure what you mean. Can you show me an example?

So in toki pona, there are usually many English alternatives, which take a while/some thinking to enumerate, and few toki pona alternatives.

To create more exercises with minimal work, I would take an exercise like this:

```yaml
- Phrase: ona li pali
  Translation: They work
  Alternative versions:
    - tenpo [ni/lon] la, [ona/ona mute] li pali
    - [ona/ona mute] li pali lon tenpo [ni/lon]
  Alternative translations:
    - [He/She/It] [works/is working/will work]
    - They [work/are working/will work]
    - It is [work/a job]
    - They are [work/jobs]
```

Copy and paste it, change the `Translation` and `Alternative versions` only (which is much easier than writing a phrase from scratch, because I do not have to think about `Alternative translations`, which is complex -- even in the above, there are translations missing). For example:

```yaml
- Phrase: ona li pali
  Translation: She worked
  Alternative versions:
    - tenpo pini la, ona li pali
    - ona li pali lon tenpo pini
  Alternative translations:
    - [He/She/It] [works/is working/will work]
    - They [work/are working/will work]
    - It is [work/a job]
    - They are [work/jobs]
```

Sure, it's not as good as a whole new phrase, but to me it is better than not having anything at all.

**Comment by zmbc** [comment]
> Another thing that we can do is pre-parse the template syntax and send it as a JSON structure.

Yes, I actually think this will be a necessity to get any kind of performant solution to this problem -- we will have to do some computation at compilation time and some on the frontend when the user submits their answer. Whatever the balance is, the JSON file should have some kind of data structure (not a simple array of string options, and not just the template string). I have some ideas, but I need to work on them a bit more before they are ready for me to actually propose something.

> One thing I can imagine is that we generate all possible solutions on the fly as a generator, so never store all of them in memory

So we could try to implement something like this and measure real-world performance, but I have a very strong hunch that anything that would iterate 400k times in the above example is not going to work well. I think we will almost certainly need to have an algorithm that does not enumerate all possibilities. One of the reasons for my hunch is that I know for a fact that Duolingo has done just that, and I don't think they would have gone to the trouble if it wasn't necessary.

As you said, it is a parsing problem, but in a very restricted grammar space. We will need to make some decisions about just how restricted -- we've already discussed one, which is whether to allow nested options (Duolingo does not). There are other restrictions to consider, such as whether some languages (likely the ones that use alphabets) can only have options at word boundaries -- Duolingo does have this distinction, and I assume they did it for a reason.

The more we restrict the template system, though, the more likely it is that some sentences will require many *templates* -- which then introduces the non-trivial problem of how to efficiently search for the templates that have a chance of matching. In Duolingo's system, I have no idea how efficiently they use their templating system (perhaps they store many alternatives as separate strings which could be condensed into a single template) but apparently some real-world sentences have thousands or tens of thousands of *templates*.

Like I said, I will think more about this and maybe come up with some rough proposals or approaches we could take.

**Comment by kantord** [comment]
> As you said, it is a parsing problem, but in a very restricted grammar space. 

But, can parsing model also approximate results? For example, can a parser be built in such a way that it parses anything that's _close enough_?

**Comment by zmbc** [comment]
> But, can parsing model also approximate results?

Oops yes, I forgot to mention that this is the other part that is tricky. I definitely think it can be done, but it does make it a bit different than a standard parsing problem. I certainly do not think we can use off-the-shelf parsing libraries to do this.

By the way, as an addendum to what I said earlier:

> My example template above would take many templates to express if there was no nesting allowed.

The example toki pona template would require 24 templates if not for nesting, which is lower than I initially thought. It does make me wonder why Duolingo has sentences with so many...

**Comment by zmbc** [comment]
Thanks for sharing that.

I agree that most sentences will likely have a very small number of templates, but it is something we need to think about -- even if only a few sentences per course need hundreds, if submitting responses for those sentences took 5 seconds to verify, that would be a serious usability problem. Of course I made up that 5 second number, so we'd have to actually measure.

My biggest question is: why are there Duolingo sentences with hundreds or thousands of options? I see at least two possibilities:

- Duolingo encourages course authors to use templates, but later user submissions are included as simple strings. Many of these could be compressed into templates, but they aren't, leading to the thousands of entries.
- There are really some languages where the Duolingo template syntax (similar to what we've been discussing, with no nesting) often **requires** thousands of templates.

This would be very useful to know the answer to! If it's the former, there are a number of pretty straightforward approaches we could take. If it's the latter, we should determine what would need to be added to the templating syntax to reduce the number of templates, or figure out what optimizations we can use in those cases to quickly find matching templates. (Those two methods might sound different, but I have a feeling they would conceptually be very similar.)

I think the contractions support is a nice-to-have, but it doesn't affect the underlying algorithm, so that could be addressed later.

**Comment by chickendude** [comment]
Where have you seen that there are thousands of templates? My guess is that the earliest courses (such as Spanish and French) may have had that issue, but most of those courses have been rewritten so that's probably not the case now. I would think in the majority of cases it's:
1. The sentences could be compressed into templates, but it gets harder to manage once you've got 15-20 so new sentences were added.
2. Edge cases (where are you going? -> whither goest thou?)
3. Overly ambiguous sentences, e.g. in your example, i think it may be better to split them up into two separate sentences to avoid all the translations. Longer sentences are harder and more annoying as well outside of reading/recognition. But even there i'd think 30-40 template sentences could easily handle 99% of reasonable answers.

I think a sentence needing hundreds of templates is going to quickly become unmaintainable. I think it would be better to look into ways to prevent sentences from needing so many separate template sentences to begin with.

But yeah, regardless of the template system, figuring out a way to go through over 400k potential answers is going to require some thinking. I don't think most sentences will have near that many, but i don't doubt that some sentences especially in more ambiguous/context-dependent languages like Toki Pona.

**Comment by zmbc** [comment]
> Where have you seen that there are thousands of templates?

https://forum.duolingo.com/comment/38731989/New-browser-extension-for-viewing-the-lists-of-accepted-solutions -- there are people reporting 3,000, and one person even said they saw a sentence with 20,000. A little hard to verify and unfortunately this stuff is not published anywhere.

I did a little more digging into the source of that extension, though, and I'm pretty sure Duolingo does not actually store them all separately. That is, even if *course authors* cannot nest brackets, the actual data structure they use to check algorithms supports a deeply nested approach. In fact, it looks like all templates are ultimately "compiled" into a single grammar-like structure.

> in your example, i think it may be better to split them up into two separate sentences to avoid all the translations

Maybe in many cases that would work well (there's certainly no need for a long sentence to teach vocab, for instance). But in Toki Pona, the method above is the only way of having a multi-clause sentence, and absolutely will need to be taught somewhere in the course.

All in all, it seems to me that Toki Pona is a particularly challenging case for alternatives, but it's not totally out of the league of other languages. [According to this Duolingo user](https://forum.duolingo.com/comment/46398772?comment_id=46402157), there is a sentence in the Norwegian course with at least 195,000 translations. For that reason, I feel like it's good both for my course and for future courses that whatever system we create support translation sets on that scale.

**Comment by zmbc** [comment]
> I think a sentence needing hundreds of templates is going to quickly become unmaintainable.

I definitely agree, and that's part of why I like nesting. For Toki Pona, anyway, I would only need 1 or 2 nested templates. Of course, if we supported nesting, course authors who found it confusing wouldn't have to use it. But it would definitely make my life a lot easier. If we end up not including nesting, I'll almost certainly make a script to implement it as a preprocessor, because having to make an edit to a template that is spread out over 24 combinations would not be fun.

**Comment by zmbc** [comment]
By the way @kantord I didn't explicitly ask, but my hope with the above proposal is to get your feedback and settle on a design before starting implementation of it. Do you have any concerns about the described approach?

**Comment by kantord** [comment]
> Does this imply that they could be out of order? If so, we'd start to see O(N^2 * M) performance since each lookup would require us to iterate.

This is true, but we can still convert them into a hash map (object).

An alternative and perhaps better idea is to use a JSON object in the first place, with the keys as IDs. What do you think? I know it's a little bit more wasteful in network terms of course

**Comment by kantord** [comment]
> In order to make things simple and allow all the optimizations above to be used across the full tree, there should only be one start node and one end node.

>  We could explicitly mark the start and end node with a special field e.g. "start": true but I think regardless the start node should always be the first item and the end the last, for the lookup reason mentioned above.

Yes, or if we use a JSON object instead of an array, then the IDs could be `"start"` and `"end"` respectively

**Comment by kantord** [comment]
> But I think the small performance gain is probably not worth the complexity.

I agree with this, it already sounds quite complex 😅

**Comment by zmbc** [comment]
> Yes, or if we use a JSON object instead of an array, then the IDs could be "start" and "end" respectively

This is a great point; I agree a JSON object is a cleaner approach. Slightly more wasteful data-wise as you said but it's a small difference.

**Comment by kantord** [comment]
@zmbc great!

can you think of other major questions to be resolved?

**Comment by zmbc** [comment]
No, I think if the design above sounds good to you, I could start implementation. I'll also plan to create some kind of technical documentation about the feature, based on the above.

**Comment by kantord** [comment]
> No, I think if the design above sounds good to you, I could start implementation. I'll also plan to create some kind of technical documentation about the feature, based on the above.

Amazing!!!!! It sounds good to me, yes

**Comment by JoelNiemela** [comment]
This sounds interesting, can I work on this?

**Comment by kantord** [comment]
> This sounds interesting, can I work on this?

sure!

**Comment by giorginogreg** [comment]
Are there any updates on this? @JoelNiemela

**Comment by giorginogreg** [comment]
It is still not really clear for me what kind of problematics **in the app** does this causes. In which exercises are there any problems?
I tried to context the (for me) improvement for each type of exercise, but in no one i saw any improvements. Could you please make some example with some exercises in the app so that i could better understand and work also on this?

**Comment by chickendude** [comment]
This just makes it easier to create alternative solutions. For example:
**They are learning**
ellos están aprendiendo
están aprendiendo
ellas están aprendiendo
están aprendiendo ellos
están aprendiendo ellas

Could be shortened into:
[ellos/ellas/] están aprendiendo.
Están aprendiendo [ellos/ellas].
